[libpas] Implement Retag-on-Scavenge for segregated objects

Achierius opened the pull request:
#### 0de9a179e544a6932f063d2b00c6781de4b5700a
<pre>
[libpas] Implement Retag-on-Scavenge for segregated objects
<a href="https://bugs.webkit.org/show_bug.cgi?id=302663">https://bugs.webkit.org/show_bug.cgi?id=302663</a>
<a href="https://rdar.apple.com/164906028">rdar://164906028</a>

Reviewed by Dan Hecht.

The original implementation of retag-on-scavenge was insufficient,
but at the time its problems were masked by other bugs (mainly
the performance bug wherein we allocated all pages with backing
MTE memory, even when they were intended only for allocating
compact objects). Secondly, they were also masked by recently-
modified page configurations which ensured that certain allocation
paths were not reached in practice in those places where the feature
was enabled. This patch goes over and fixes those issues for segregated
objects in particular, thus enabling the feature in privileged
processes.
The main issues:
 - The way we were previously checking whether or not an allocation
   should be re- tagged relied on the page configuration. However, that
   does not always work because the local-allocator can also decide to
   use a tagged page-config to ultimately allocate *untagged* memory.
   Since the metadata used to make this decision is not accessible on
   the deallocation path, we instead determine it &apos;experimentally&apos; by
   executing an LDG instruction on the page. Since libpas makes sure to
   never explicitly zero-tag an allocation inside of an MTE-backed page,
   if we get a zero-tag back that means the object is canonically tagged
   -- i.e., has no backing MTE tag memory.
 - Bitfit heaps, unlike segregated heaps, can allocate objects of
   different sizes in the same memory location. This complicates both
   the process of determining the size of an object being freed as well
   as making it impossible to elide re-tagging an object on allocation,
   as the deallocation path cannot know the size of the next object to
   be allocated in a given slot. This patch thus disables the bitfit
   retag-on-scavenge path for the time being, pending future work.

As part of the above work, this patch also refactors some components to
use always-inline functions instead of macros, as part of a general
intent to move away from the heavily macro-ified and indirect approach
that was an artifact of how libpas MTE was developed on a side-tree prior
to the announcement of MIE.

* Source/bmalloc/libpas/src/libpas/pas_bitfit_page_inlines.h:
(pas_bitfit_page_finish_allocation):
* Source/bmalloc/libpas/src/libpas/pas_fast_megapage_table.h:
* Source/bmalloc/libpas/src/libpas/pas_local_allocator_inlines.h:
(pas_local_allocator_try_allocate_in_primordial_partial_view):
(pas_local_allocator_try_allocate_with_free_bits):
* Source/bmalloc/libpas/src/libpas/pas_mte.h:
(pas_mte_maybe_tag_allocated_region):
(pas_mte_retag_freed_region_if_tagged):
* Source/bmalloc/libpas/src/libpas/pas_mte_config.h:
* Source/bmalloc/libpas/src/libpas/pas_page_granule_use_count.h:
* Source/bmalloc/libpas/src/libpas/pas_zero_memory.h:
(pas_zero_memory):

Canonical link: <a href="https://commits.webkit.org/303396@main">https://commits.webkit.org/303396@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2857f697563a67c6a4f4c8beb5fe43d7bfe1e181

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132173 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4666 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43198 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139688 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/84094 "Built successfully") | ⏳ 🛠 ios-apple 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4700 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4427 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101031 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/68365 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/671e3d07-994e-4fef-ace6-8cfed0802f99) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135119 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/3327 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118384 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81824 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/0c0938ee-6eaf-4b1e-a421-1b547730830a) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/3211 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/1065 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82908 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/124236 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/112031 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36504 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142335 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/130680 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4335 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37083 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109408 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4416 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/3762 "Found 1 new test failure: imported/w3c/web-platform-tests/webvtt/rendering/cues-with-video/processing-model/selectors/cue/background_properties.html (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109586 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27784 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3290 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114656 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/57581 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4389 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33043 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/163647 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4221 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/42528 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/4480 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4348 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_mte_config.h:
```diff
@@ -122,7 +122,7 @@ extern Slot g_config[];
 
 #define PAS_MTE_FEATURE_FORCED(feature) (0)
-#define PAS_MTE_FEATURE_HARDENED_FORCED(feature) (feature == PAS_MTE_FEATURE_ADJACENT_TAG_EXCLUSION)
+#define PAS_MTE_FEATURE_HARDENED_FORCED(feature) (feature == PAS_MTE_FEATURE_RETAG_ON_SCAVENGE || feature == PAS_MTE_FEATURE_ADJACENT_TAG_EXCLUSION)
```
what does "forced" mean in this context? And hardened? Maybe these two lines are saying what MTE "features" are enabled for non-hardened and hardened processes? The "forced" seems confusing, in what sense is it forced?

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_mte.h:
```diff
@@ -201,11 +174,23 @@ PAS_IGNORE_WARNINGS_BEGIN("unsafe-buffer-usage")
 // objects allocated by the allocator are the same size, e.g. like is the
 // case with any slab allocator.
-enum pas_mte_allocator_homogeneity {
+enum pas_allocator_homogeneity {
```
Is it no longer pte specific?

Achierius replied:
Yeah, it's doesn't strictly have to do with MTE but rather is a property of the allocators itself.

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_mte.h:
```diff
@@ -201,11 +174,23 @@ PAS_IGNORE_WARNINGS_BEGIN("unsafe-buffer-usage")
 };
 
+typedef enum pas_allocator_homogeneity pas_allocator_homogeneity;
+typedef enum pas_allocator_homogeneity __pas_allocator_homogeneity;
```
why do we have these __ variants?

Achierius replied:
Vaguely holding to general libpas practice but there's no real point here as far as I understand, so I've removed them for this patch

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_mte.h:
other or adjacent?

dhecht replied:
who is "us"? does that mean "libpas"?

Achierius replied:
Changed to `libpas`

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_mte.h:
I don't follow this sentence.

dhecht replied:
I think it's saying something like: once a this region has a valid tag, we can relax ...

Achierius replied:
Reworded

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_mte.h:
it does change delay when UAF is caught for a particular deallocation. so not sure exactly what "nothing" means.

Achierius replied:
Reworded

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_mte.h:
Did you mean to keep these printfs?

dhecht replied:
If so, why not pas_log?

Achierius replied:
Because we already use printf elsewhere in the file -- I'll be moving them all to pas_log at one time

Achierius commented on Source/bmalloc/libpas/src/libpas/pas_mte.h:
```diff
@@ -141,37 +145,6 @@ PAS_IGNORE_WARNINGS_BEGIN("unsafe-buffer-usage")
             : \
         ); \
     } while (0)
-#define PAS_MTE_CHECK_TAG_AND_SET_TCO(ptr) do { \
```
This part had to be moved to pas_mte_config.h so that it could be available to pas_zero_memory.h without making the latter include pas_mte.h.

Achierius commented on Source/bmalloc/libpas/src/libpas/pas_mte.h:
```diff
@@ -464,6 +449,38 @@ PAS_IGNORE_WARNINGS_END
         b &= ~PAS_MTE_TAG_MASK; \
     } while (0)
 
+// Use these to configure the tagging policy for different sizes. Currently we only
```
This whole chunk was just moved upwards from later on in the file, as it's now needed by pas_mte_tag_allocated_region_as_appropriate, a little further down in the file.

Achierius commented on Source/bmalloc/libpas/src/libpas/pas_mte.h:
```diff
@@ -496,14 +513,106 @@ PAS_IGNORE_WARNINGS_END
         } \
     } while (0)
 
-#define PAS_MTE_TAG_REGION_FROM_INITIAL_ALLOCATION(ptr, size, mode, is_allocator_homogeneous, is_known_medium) do { \
```
This, as well as PAS_MTE_TAG_REGION_FROM_OTHER_ALLOCATION, were combined into pas_mte_tag_allocated_region_as_appropriate -- the `initiality` parameter distinguishes between what was previously the _INITIAL_ path / the _OTHER_ path.

Achierius commented on Source/bmalloc/libpas/src/libpas/pas_mte.h:
This replaces PAS_MTE_TAG_REGION_FROM_DEALLOCATION, and is where most of the changes are located.
The big difference in the logic is that we ldg the allocated memory to see whether or not it has a tag, and only if it does do we retag it -- the previous logic was faulty in that it would end up tagging allocations that were not allocated inside MTE memory.

Achierius commented on Source/bmalloc/libpas/src/libpas/pas_mte.h:
The main difference in this function (vs. the prior logic in PAS_MTE_TAG_REGION_FROM_INITIAL_ALLOCATION/PAS_MTE_TAG_REGION_FROM_OTHER_ALLOCATION) is that we distinguish "maybe_initial" allocations from "non_initial". This is important for bitfit allocations, where practically we can't really be sure whether the allocation was the first to be allocated in all of its bytes -- so essentially we can't make any assumptions about it, either that it was non-initial and thus we can skip alloc-tagging, nor that it was definitely-initial and thus we can assume the previous tag was 0.

Achierius commented on Source/bmalloc/libpas/src/libpas/pas_mte.h:
```diff
@@ -552,16 +639,6 @@ PAS_IGNORE_WARNINGS_END
 
 #if PAS_OS(DARWIN)
 
-// We can't check whether PAS_ASSERT is defined since this header is included early
```
This chunk was removed because it was only used by PAS_MTE_ZERO_FILL_PAGE, but now that we have pas_utils.h as an include we can just use PAS_ASSERT there instead.

Achierius commented on Source/bmalloc/libpas/src/libpas/pas_mte.h:
```diff
@@ -603,19 +680,6 @@ PAS_IGNORE_WARNINGS_END
 // bit in its page-table entries.
 #define PAS_MTE_HANDLE_ZERO_FILL_PAGE(ptr, size, flags, tag) PAS_MTE_ZERO_FILL_PAGE(ptr, size, flags, tag)
 
-// Used to allow us to toggle TCO before setting large chunks of
```
This was removed because pas_zero_memory.h can't include this header, so it just handles the PAS_USE_MTE logic itself as a special case.

dhecht replied:
Do we plan to move away from this PAS_MTE_HANDLE pattern eventually or is the plan to keep it around?

Achierius replied:
Definitely planning to remove it, it's just an artifact of how we previously had to develop on a side-branch &c

Achierius commented on Source/bmalloc/libpas/src/libpas/pas_mte.h:
```diff
@@ -686,38 +750,6 @@ PAS_IGNORE_WARNINGS_END
 // Used to clear pointer tag bits when summarizing a range in the large sharing pool.
 #define PAS_MTE_HANDLE_LARGE_SHARING_POOL_COMPUTE_SUMMARY(a, b) PAS_MTE_CLEAR_PAIR(a, b)
 
-// Use these to configure the tagging policy for different sizes. Currently we only
```
This whole chunk was just moved up higher into the file because one of the functions above needs it.

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_mte_config.h:
```diff
@@ -140,10 +140,49 @@ extern Slot g_config[];
+            : "memory" \
+        ); \
+    } while (0)
+#define PAS_MTE_SET_TCO_UNCHECKED do { \
```
nit: a newline between the defines would help readability

Achierius replied:
Done

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_page_granule_use_count.h:
```diff
@@ -47,6 +47,7 @@ static PAS_ALWAYS_INLINE void pas_page_granule_get_indices(
     PAS_ASSERT(*index_of_last_granule < page_size / granule_size);
 }
 
+PAS_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
```
why is this needed in this change?

Achierius replied:
Some obscure case of the compiler being able to infer something is unsafe-pointer-arithmetic where it wasn't before.

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_mte.h:
```diff
@@ -65,7 +69,7 @@ PAS_IGNORE_WARNINGS_BEGIN("unsafe-buffer-usage")
 #define PAS_MTE_SMALL_PAGE_NO(ptr) (((uintptr_t)ptr) & PAS_MTE_SMALL_PAGE_NO_MASK)
 
-#define PAS_MTE_GET_TAG(ptr) do { \
+#define PAS_MTE_GET_MTAG(ptr) do { \
```
What does the M stand for? Memory? Distinguishing it from what? I think ARM calls it "Allocation tag", why not stick to that termination? And why doesn't PAS_MTE_SET_TAG have the M?

dhecht replied:
As we discussed, if you do change more to MTAG, let's do that separately as to not pollute this diff more.

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_mte.h:
I think you should explicitly state the punch-line: And so, we use a Retag-on-Free policy for homogeneous pages (or whatever the terminology is).

dhecht replied:
Or I guess more accurately, which policy is used depending on the feature flag setting.

dhecht replied:
And depending on bitfit vs segregated i suppose

Achierius replied:
Added more diagrams / explanation with specifics

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_mte.h:
why printf and not pas_log?

dhecht replied:
Also, is there a reason why the logging here happens after the action, but above the logging occurs before MTE_PURIFY? If unnecessary, the inconsistency just adds some friction to reading the code.

Achierius replied:
Fixed logging order; talked about printf in earlier comment

dhecht replied:
Order not fixed

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_mte.h:
I find the negatives hard to reason about: how about known_already_tagged? That's the information you want to infer, I think?

Achierius replied:
Refactored to have less conditional depth

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_mte.h:
I think the "as_appropriate" doesn't add much info and is distracting. Maybe pas_mte_maybe_tag_allocated_region to indicate the region may or may not end up being tagged?

Achierius replied:
Changed

dhecht replied:
Name doesn't looked changed.

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_mte.h:
Maybe call it: pas_mte_retag_freed_region_if_tagged

Achierius replied:
Changed

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_mte.h:
I don't understand why we tag here. I thought the point is to avoid tagging on allocation if it was tagged during scavenge and it's not the initial allocation time.

Achierius replied:
This is for the non-compact path, which will naturally be '0-tagged' (in actuality not actually tagged at all)

dhecht commented on Source/bmalloc/libpas/src/libpas/pas_mte.h:
I think it would be clearer to do the feature check first since the known_non_initial check is relevant to that feature and so it's clearer if that part is guarded as well.

Achierius replied:
Refactored this to have less conditional depth

dhecht replied:
Code looks the same.

Achierius committed:
[libpas] Implement Retag-on-Scavenge for segregated objects
https://bugs.webkit.org/show_bug.cgi?id=302663
rdar://164906028

Reviewed by Dan Hecht.

The original implementation of retag-on-scavenge was insufficient,
but at the time its problems were masked by other bugs (mainly
the performance bug wherein we allocated all pages with backing
MTE memory, even when they were intended only for allocating
compact objects). Secondly, they were also masked by recently-
modified page configurations which ensured that certain allocation
paths were not reached in practice in those places where the feature
was enabled. This patch goes over and fixes those issues for segregated
objects in particular, thus enabling the feature in privileged
processes.
The main issues:
 - The way we were previously checking whether or not an allocation
   should be re- tagged relied on the page configuration. However, that
   does not always work because the local-allocator can also decide to
   use a tagged page-config to ultimately allocate *untagged* memory.
   Since the metadata used to make this decision is not accessible on
   the deallocation path, we instead determine it 'experimentally' by
   executing an LDG instruction on the page. Since libpas makes sure to
   never explicitly zero-tag an allocation inside of an MTE-backed page,
   if we get a zero-tag back that means the object is canonically tagged
   -- i.e., has no backing MTE tag memory.
 - Bitfit heaps, unlike segregated heaps, can allocate objects of
   different sizes in the same memory location. This complicates both
   the process of determining the size of an object being freed as well
   as making it impossible to elide re-tagging an object on allocation,
   as the deallocation path cannot know the size of the next object to
   be allocated in a given slot. This patch thus disables the bitfit
   retag-on-scavenge path for the time being, pending future work.

As part of the above work, this patch also refactors some components to
use always-inline functions instead of macros, as part of a general
intent to move away from the heavily macro-ified and indirect approach
that was an artifact of how libpas MTE was developed on a side-tree prior
to the announcement of MIE.

* Source/bmalloc/libpas/src/libpas/pas_bitfit_page_inlines.h:
(pas_bitfit_page_finish_allocation):
* Source/bmalloc/libpas/src/libpas/pas_fast_megapage_table.h:
* Source/bmalloc/libpas/src/libpas/pas_local_allocator_inlines.h:
(pas_local_allocator_try_allocate_in_primordial_partial_view):
(pas_local_allocator_try_allocate_with_free_bits):
* Source/bmalloc/libpas/src/libpas/pas_mte.h:
(pas_mte_maybe_tag_allocated_region):
(pas_mte_retag_freed_region_if_tagged):
* Source/bmalloc/libpas/src/libpas/pas_mte_config.h:
* Source/bmalloc/libpas/src/libpas/pas_page_granule_use_count.h:
* Source/bmalloc/libpas/src/libpas/pas_zero_memory.h:
(pas_zero_memory):

Canonical link: https://commits.webkit.org/303396@main
